libp2p/Session.cpp

```diff
@@ -248,6 +255,24 @@ void Session::write()
             DEV_GUARDED(x_framing)
             {
                 m_writeQueue.pop_front();
+                if(m_writeQueue.empty() && disconnectImminent)
+                {
+                    ba::io_context ioc;
+                    ba::steady_timer disconnectTimer(ioc);
+                    disconnectTimer.expires_from_now(chrono::seconds(2));
+                    disconnectTimer.async_wait([this](boost::system::error_code const& ) {});
```

Comment:
You'll want to close the socket in the timer handler (since we want to close the socket after the timer has fired)...your code currently kicks off the timer (async) wait but then immediately closes the socket (if it's already open).

Note that the timer wait is asynchronous, not synchronous, which means that async_wait call returns immediately (the wait operation is inserted into a boost queue and the handler is inserted into the boost handler queue, and both queues are processed in response to calls of io_context::run on whatever thread run() is called on which in this case is the network thread

Reply:
Oh that's what the anomaly was in my testing before and after changes. So stupid of me.

Reply:
> Oh that's what the anomaly was in my testing before and after changes. So stupid of me.

@twinstar26 No worries accidents happen 😄